server/src/main/java/org/apache/lucene/index/BaseStarTreeBuilder.java

```diff
@@ -0,0 +1,648 @@
+                } catch (Exception e) {
+                    logger.error("Cannot get value for aggregation", e);
+                    throw new IllegalStateException("Cannot get value for aggregation[" + starTreeDocument.metrics[i] + "]");
+                }
+            }
+            return new StarTreeDocument(dimensions, metrics);
+        } else {
+            for (int i = 0; i < numMetrics; i++) {
+                try {
+                    aggregatedDocument.metrics[i] = metricAggregatorInfos.get(i)
+                        .getValueAggregators()
+                        .mergeAggregatedValues(starTreeDocument.metrics[i], aggregatedDocument.metrics[i]);
+                } catch (Exception e) {
+                    logger.error("Cannot apply value to aggregated document for aggregation", e);
+                    throw new IllegalStateException(
+                        "Cannot apply value to aggregated document for aggregation [" + starTreeDocument.metrics[i] + "]"
+                    );
+                }
+            }
+            return aggregatedDocument;
+        }
+    }
+
+    /**
+     * Builds the star tree using total segment documents
+     *
+     * @throws IOException when we are unable to build star-tree
+     */
+    public void build() throws IOException {
+        long startTime = System.currentTimeMillis();
+        logger.debug("Star-tree build is a go with star tree field {}", starTreeField.getName());
+
+        Iterator<StarTreeDocument> starTreeDocumentIterator = sortAndAggregateStarTreeDocuments(totalSegmentDocs);
+        logger.debug("Sorting and aggregating star-tree in ms : {}", (System.currentTimeMillis() - startTime));
+        build(starTreeDocumentIterator);
+        logger.debug("Finished Building star-tree in ms : {}", (System.currentTimeMillis() - startTime));
+    }
+
+    /**
+     * Builds the star tree using Star-Tree Document
+     *
+     * @param starTreeDocumentIterator contains the sorted and aggregated documents
+     * @throws IOException when we are unable to build star-tree
+     */
+    public void build(Iterator<StarTreeDocument> starTreeDocumentIterator) throws IOException {
+        int numSegmentStarTreeDocument = totalSegmentDocs;
+
+        while (starTreeDocumentIterator.hasNext()) {
+            appendToStarTree(starTreeDocumentIterator.next());
+        }
+        int numStarTreeDocument = numStarTreeDocs;
+        logger.debug("Generated star tree docs : [{}] from segment docs : [{}]", numStarTreeDocument, numSegmentStarTreeDocument);
+
+        if (numStarTreeDocs == 0) {
+            // TODO: Uncomment when segment codec is ready
+            // StarTreeBuilderUtils.serializeTree(indexOutput, rootNode, dimensionsSplitOrder, numNodes);
+            return;
+        }
+
+        constructStarTree(rootNode, 0, numStarTreeDocs);
+        int numStarTreeDocumentUnderStarNode = numStarTreeDocs - numStarTreeDocument;
```

Comment:
Wouldn't `numStarTreeDocumentUnderStarNode ` be always 0?

Reply:
During creation of star nodes, we do ```appendToStarTree``` which increments the value for the documents under star node.